server.go

```diff
@@ -140,10 +173,13 @@ func (u *Upgrader) Upgrade(w http.ResponseWriter, r *http.Request, responseHeade
 
 	challengeKey := r.Header.Get("Sec-Websocket-Key")
 	if challengeKey == "" {
-		return u.returnError(w, r, http.StatusBadRequest, "websocket: not a websocket handshake: `Sec-WebSocket-Key' header is missing or blank")
+		return u.returnError(w, r, http.StatusBadRequest, "websocket: not a websocket handshake: 'Sec-WebSocket-Key' header is missing or blank")
 	}
 
-	subprotocol := u.selectSubprotocol(r, responseHeader)
+	subprotocol, ok := u.selectSubprotocol(r, responseHeader)
+	if !ok {
+		return u.returnError(w, r, http.StatusBadRequest, "websocket: unsupported client subprotocol")
```

Comment:
This prevents the client from handling failed protocol negotiation.  This may break existing applications.  Continue as normal on failed negotiation.